File: 1/2020.09.27/22.Generate Parentheses/java/solution.java
class Solution {

    List<String> res = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        generate(0, 0, n, "");
        return res;
    }

    private void generate(int left, int right, int n, String s) {
        if (left == n && right == n) {
            res.add(s);
            return;
        }
        if (left < n) {
            generate(left+1, right, n, s + "(");
        }
        if (right < left) {
            generate(left, right+1, n, s + ")");
        }

    }
}

File: 1/2020.09.27/22.Generate Parentheses/python/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return []
        self.res = []
        self.aux(0, 0, n, '')
        return self.res

    def aux(self, left, right, n, s):
        if left == n and right == n:
            self.res.append(s)
        if left < n:
            self.aux(left+1, right, n, s + '(')
        if right < left:
            self.aux(left, right+1, n, s + ')')
